test/widgets/compose_box_test.dart

```diff
@@ -573,97 +600,199 @@ void main() {
-        checkComposeBox(isShown: false);
+    group('in channel/topic narrow according to channel post policy', () {
+      void checkComposeBox({required bool isShown}) => checkComposeBoxIsShown(isShown,
+        bannerLabel: zulipLocalizations.errorBannerCannotPostInChannelLabel);
+
+      final narrowTestCases = [
+        ('channel', const ChannelNarrow(1)),
+        ('topic',   const TopicNarrow(1, 'topic')),
+      ];
+
+      for (final (String narrowType, Narrow narrow) in narrowTestCases) {
+        testWidgets('compose box is shown in $narrowType narrow', (tester) async {
+          await prepareComposeBox(tester,
+            narrow: narrow,
+            selfUser: eg.user(role: UserRole.administrator),
+            streams: [eg.stream(streamId: 1,
+              channelPostPolicy: ChannelPostPolicy.moderators)]);
+          checkComposeBox(isShown: true);
+        });
+
+        testWidgets('error banner is shown in $narrowType narrow', (tester) async {
+          await prepareComposeBox(tester,
+            narrow: narrow,
+            selfUser: eg.user(role: UserRole.moderator),
+            streams: [eg.stream(streamId: 1,
+              channelPostPolicy: ChannelPostPolicy.administrators)]);
+          checkComposeBox(isShown: false);
+        });
+      }
 
-        await changeUserStatus(tester, user: deactivatedUser, isActive: true);
+      Future<void> changeUserRole(WidgetTester tester, {
+        required User user,
+        required UserRole role,
+      }) async {
+        await store.handleEvent(RealmUserUpdateEvent(id: 1,
+          userId: user.userId, role: role));
+        await tester.pump();
+      }
+
+      Future<void> changeChannelPolicy(WidgetTester tester, {
+        required ZulipStream channel,
+        required ChannelPostPolicy policy,
+      }) async {
+        await store.handleEvent(eg.channelUpdateEvent(channel,
+          property: ChannelPropertyName.channelPostPolicy, value: policy));
+        await tester.pump();
+      }
+
+      testWidgets('user role decreases -> compose box is replaced with the banner', (tester) async {
+        final selfUser = eg.user(role: UserRole.administrator);
+        await prepareComposeBox(tester,
+          narrow: const ChannelNarrow(1),
+          selfUser: selfUser,
+          streams: [eg.stream(streamId: 1, channelPostPolicy: ChannelPostPolicy.administrators)]);
         checkComposeBox(isShown: true);
+
+        await changeUserRole(tester, user: selfUser, role: UserRole.moderator);
+        checkComposeBox(isShown: false);
       });
-    });
 
-    group('group DMs', () {
-      testWidgets('compose box replaced with a banner', (tester) async {
-        final deactivatedUsers = [eg.user(isActive: false), eg.user(isActive: false)];
-        await prepareComposeBox(tester, narrow: groupDmNarrowWith(deactivatedUsers),
-          users: deactivatedUsers);
+      testWidgets('user role increases -> banner is replaced with the compose box', (tester) async {
```

Comment:
You've already taken the trouble to write these four test cases so we may as well include them here, but for future reference: I typically wouldn't ask for tests to check that updating the data in the store causes the widgets to properly update for the new values. In general we can just rely on the assumption that Flutter's rebuild mechanisms work correctly :slightly_smiling_face: 

The situation where I might want tests for widgets getting updated is when we have a stateful widget with some nontrivial logic of its own for arranging for updates to happen, in which case that logic needs tests. But here it's just `PerAccountStoreWidget.of` that causes rebuilds.

(We also rely on our `handleEvent` implementation calling `notifyListeners` in all the right situations. That isn't currently as thoroughly tested as it ideally would be — instead, for these two event types in particular and some others, we rely on the fact that the implementation straightforwardly says `notifyListeners();` at the end. But if or when at some point we do test that point, the tests in `test/model/` that exercise handling events are the right layer for that.)